hugr-llvm/src/extension/collections/array.rs

```diff
@@ -29,98 +29,154 @@ use crate::{CodegenExtension, CodegenExtsBuilder};
-pub trait VArrayCodegen: Clone {
-    /// Return the llvm type of [hugr_core::std_extensions::collections::value_array::VALUE_ARRAY_TYPENAME].
+pub trait ArrayCodegen: Clone {
+    fn emit_malloc<'c, H: HugrView<Node = Node>>(
+        &self,
+        ctx: &mut EmitFuncContext<'c, '_, H>,
+        size: IntValue<'c>,
+    ) -> Result<BasicValueEnum<'c>> {
+        emit_libc_malloc(ctx, size.into())
+    }
+
+    fn emit_free<'c, H: HugrView<Node = Node>>(
+        &self,
+        ctx: &mut EmitFuncContext<'c, '_, H>,
+        ptr: PointerValue<'c>,
+    ) -> Result<()> {
+        emit_libc_free(ctx, ptr.into())
+    }
+
+    /// Return the llvm type of [hugr_core::std_extensions::collections::array::ARRAY_TYPENAME].
     fn array_type<'c>(
         &self,
-        _session: &TypingSession<'c, '_>,
+        session: &TypingSession<'c, '_>,
         elem_ty: BasicTypeEnum<'c>,
-        size: u64,
+        _size: u64,
     ) -> impl BasicType<'c> {
-        elem_ty.array_type(size as u32)
+        array_struct_ty(session, elem_ty)
     }
 
-    /// Emit a [hugr_core::std_extensions::collections::value_array::VArrayValue].
+    /// Emit a [hugr_core::std_extensions::collections::array::ArrayValue].
     fn emit_array_value<'c, H: HugrView<Node = Node>>(
         &self,
         ctx: &mut EmitFuncContext<'c, '_, H>,
-        value: &value_array::VArrayValue,
+        value: &array::ArrayValue,
     ) -> Result<BasicValueEnum<'c>> {
         emit_array_value(self, ctx, value)
     }
 
-    /// Emit a [hugr_core::std_extensions::collections::value_array::VArrayOp].
+    /// Emit a [hugr_core::std_extensions::collections::array::ArrayOp].
     fn emit_array_op<'c, H: HugrView<Node = Node>>(
         &self,
         ctx: &mut EmitFuncContext<'c, '_, H>,
-        op: VArrayOp,
+        op: ArrayOp,
         inputs: Vec<BasicValueEnum<'c>>,
         outputs: RowPromise<'c>,
     ) -> Result<()> {
         emit_array_op(self, ctx, op, inputs, outputs)
     }
 
-    /// Emit a [hugr_core::std_extensions::collections::value_array::VArrayRepeat] op.
+    fn emit_array_clone<'c, H: HugrView<Node = Node>>(
+        &self,
+        ctx: &mut EmitFuncContext<'c, '_, H>,
+        op: ArrayClone,
+        array_v: BasicValueEnum<'c>,
+    ) -> Result<(BasicValueEnum<'c>, BasicValueEnum<'c>)> {
+        let elem_ty = ctx.llvm_type(&op.elem_ty)?;
+        let (array_ptr, array_offset) = decompose_array_value(ctx.builder(), array_v)?;
+        let (other_ptr, other_array_v) = build_array_malloc(ctx, self, elem_ty, op.size)?;
+        let src_ptr = unsafe {
+            ctx.builder()
+                .build_in_bounds_gep(array_ptr, &[array_offset], "")?
+        };
+        let length = usize_ty(&ctx.typing_session()).const_int(op.size, false);
+        let size_value = ctx
+            .builder()
+            .build_int_mul(length, elem_ty.size_of().unwrap(), "")?;
+        emit_libc_memcpy(ctx, other_ptr.into(), src_ptr.into(), size_value.into())?;
```

Comment:
Is `memcpy` available on our platform, or do we need to make this more configurable?